API v2

SebaSOFT opened the pull request.

Commit:
- hasUserBought with customerId
- plenigoToken header NOT WORKING
- header handling to ADD and not replace headers

Commit:
JWT class

Commit:
- getProductsBought can receive a customer Id too for White labeling...

Commit:
- Checkout now accepts a logintToken parameter in the build method (fixes PLENIGO-921 )
- RestClient now supports JWT
- JWT util shouldnt have the STANDARD "typ" field
- API v02 compatibility (fixes PLENIGO-912 )
- Added a test for the checkout process with loginToken

Commit:
- JWT only supports a single algorythm (sha256)
- Added additional fields to ProductData (fixes PLENIGO-919 )

Commit:
- Tests for new fields on product data

Commit:
- JavaScript function was called: checkoutWithRemoteLogin

Commit:
- External User Management and tests
- New put requests for the SDK

Commit:
- Identation, Imports optimized and TODO taken out

Commit:
- Fixed a problem with PlenigoManager availability
- Added some tests in sake of coverage :D

maximilianschweitzer commented on src/plenigo/builders/CheckoutSnippetBuilder.php:
```diff
@@ -52,13 +50,13 @@ public function __construct(Product $productToChkOut)
+     * @param string $loginToken A string contianing the loginToken if obtained from the external user management
      *
      * @return string A Javascript snippet that is compliant with plenigo's Javascript
      * SDK.
```
Unnecessary line break

maximilianschweitzer commented on src/plenigo/internal/services/Service.php:
```diff
@@ -137,6 +137,29 @@ protected static function postJSONRequest($endPoint, $oauth = false, array $para
+        if($oauth){
+            $clazz = get_class();
+            PlenigoManager::notice($clazz, "OAUTH JSON PUT REQUEST");
+            $url = PlenigoManager::get()->getUrlOAuth() . $endPoint;
```
when is oauth json put used?

SebaSOFT replied:
Not used but available

maximilianschweitzer commented on src/plenigo/internal/utils/JWT.php:
```diff
@@ -0,0 +1,193 @@
+     */
+    public static function decode($jwt, $key = null, $verify = true) {
+        $tks = explode('.', $jwt);
+        if (count($tks) != 3) {
```
why 3

SebaSOFT replied:
a JWT token is a string consisting of 3 parts separated but dots

http://jwt.io

maximilianschweitzer replied:
Still a magic number...

maximilianschweitzer commented on src/plenigo/internal/utils/JWT.php:
```diff
@@ -0,0 +1,193 @@
+     * @uses jsonDecode
+     * @uses urlsafeB64Decode
+     */
+    public static function decode($jwt, $key = null, $verify = true) {
```
when are we decoding a jwt token client on the client side?

SebaSOFT replied:
This method is for decoding the token on the server side...

maximilianschweitzer commented on src/plenigo/internal/utils/JWT.php:
```diff
@@ -0,0 +1,193 @@
+     * @return object          Object representation of JSON string
+     * @throws \Exception Provided string was invalid JSON
+     */
+    public static function jsonDecode($input) {
```
do we need this function if we don't decode a jwt token on client side?

SebaSOFT replied:
We need this for testing, also this is PHP, so server side

maximilianschweitzer commented on src/plenigo/internal/utils/JWT.php:
```diff
@@ -0,0 +1,193 @@
+     * @return string A decoded string
+     */
+    public static function urlsafeB64Decode($input) {
+        $remainder = strlen($input) % 4;
```
every used if we don't decode a jwt token?

SebaSOFT replied:
We do it on the tests...

maximilianschweitzer replied:
Yes but perhaps we should move all of the functionality to the the test class itself. We are delivering code "for the future". Normally those code will never be used again.

maximilianschweitzer commented on src/plenigo/internal/utils/JWT.php:
```diff
@@ -0,0 +1,193 @@
+        );
+        $clazz = get_class();
+        $errMsg = isset($messages[$errno]) ? $messages[$errno] : 'Unknown JSON error: ' . $errno;
+        PlenigoManager::warning($clazz, $errMsg);
```
is this only a warning if an exception is thrown?

maximilianschweitzer commented on src/plenigo/services/UserManagementService.php:
```diff
@@ -0,0 +1,161 @@
+        if (isset($data->loginToken)) {
+            $result = $data->loginToken;
+        } else {
+            $result = "" . $data;
```
why are you returning the data string here? Cannot contain data here.

SebaSOFT replied:
The $result object just contains the loginToken property so we are returning that. If there is an error or otherwise different response, we fail gracefully by returnin a string representation of the body of the response

maximilianschweitzer commented on src/plenigo/services/UserManagementService.php:
```diff
@@ -0,0 +1,161 @@
+        if (isset($data->customerId)) {
+            $result = $data->customerId;
+        } else {
+            $result = "" . $data;
```
Same like below. Cannot really contain data here.

maximilianschweitzer commented on src/plenigo/services/UserService.php:
```diff
@@ -221,41 +207,46 @@ public static function isLoggedIn() {
+        if (is_null($pCustId)) {
+            $cookieText = static::getCookieContents(PlenigoManager::PLENIGO_USER_COOKIE_NAME);
+            if (!isset($cookieText) || is_null($cookieText) || !is_string($cookieText) || empty($cookieText)
+            ) {
```
No new line here.

maximilianschweitzer commented on src/plenigo/services/UserService.php:
```diff
@@ -221,41 +207,46 @@ public static function isLoggedIn() {
+            if (!isset($cookieText) || is_null($cookieText) || !is_string($cookieText) || empty($cookieText)
+            ) {
+                $clazz = get_class();
+                PlenigoManager::notice($clazz, "Plenigo cookie not set!!");
```
"and no customer id given" is missing

Commit:
- Fix weird cURL bug that doesnt allow PHP to send POST body in a PUT request (rendering this method non-compliant with REST)
- Even more cURL output for debugging, perhaps too much
- PHPDoc now says about receiving arrays in hasUserBought()

Commit:
- closing tempfiles that we open... avoid memory leak

Commit:
- Max corrections

Commit:
- Test before you push!

Commit:
- OAuth GONE
